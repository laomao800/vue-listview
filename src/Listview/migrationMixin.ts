import Vue from 'vue'
import get from '@/utils/getValue'
import camelCase from 'lodash/camelCase'
import { warn } from '@/utils/debug'
import { hasOwn } from '@/utils'

const insteadWarn = (type: string, oldName: string, newName: string) =>
  warn(
    `[Migration][${type}]: '${oldName}' is removed, use '${newName}' instead.`
  )

export default Vue.extend({
  mounted() {
    // Props
    const props = get(this, '$vnode.data.attrs', {})
    const migrationProps: Record<string, string> = {
      tableProps: 'contentProps',
      tableEvents: 'contentEvents',
      tableSelection: 'selection',
    }
    Object.entries(props).forEach(([propName, propValue]) => {
      propName = camelCase(propName)
      if (migrationProps[propName]) {
        insteadWarn('Props', propName, migrationProps[propName])
      }
      if (propName === 'filterFields') {
        if (
          Array.isArray(propValue) &&
          propValue.some((field: any) => hasOwn(field, 'get'))
        ) {
          insteadWarn('Props', 'filterFields.get()', 'transformRequestData()')
        }
      }
      if (propName === 'contentMessage') {
        if (hasOwn(propValue as any, 'message')) {
          insteadWarn('Props', 'contentMessage.message', 'contentMessage.text')
        }
      }
    })

    // Slots
    const migrationSlots: Record<string, string> = {
      'prepend-filterbar-submit': 'prepend-submit',
      'append-filterbar-submit': 'append-submit',
      'prepend-filterbar-more': 'prepend-more',
      'append-filterbar-more': 'append-more',
    }
    Object.keys(this.$slots).forEach((name) => {
      if (migrationSlots[name]) {
        insteadWarn('Slots', name, migrationSlots[name])
      }
    })
  },
})
