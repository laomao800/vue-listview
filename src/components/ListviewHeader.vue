<template>
  <div class="lv__header">
    <h1 v-if="headerTitle" class="lv__header-title">{{ headerTitle }}</h1>
    <el-breadcrumb
      v-if="internalNav.length > 0"
      separator="/"
      class="lv__header-breadcrumb"
    >
      <el-breadcrumb-item
        v-for="(item, index) in internalNav"
        :key="index"
        :to="item.to"
      >
        {{ item.text }}
      </el-breadcrumb-item>
    </el-breadcrumb>
  </div>
</template>

<script>
import isPlainObject from 'lodash/isPlainObject'

export default {
  name: 'ListviewHeader',

  inheritAttrs: false,

  props: {
    headerTitle: { type: String, default: '' },
    headerNav: { type: Array, default: () => [] },
  },

  computed: {
    internalNav() {
      const validNav = []
      this.headerNav.forEach((nav) => {
        let text, to
        if (typeof nav === 'string' && !!nav) {
          text = nav
        } else if (isPlainObject(nav)) {
          text = nav.text
          to = nav.to
        }
        if (text) {
          validNav.push({ text, to })
        }
      })
      return validNav
    },
  },
}
</script>

<style lang="less">
.lv__header {
  margin: -10px -10px 10px;
  padding: 8px 12px;
  line-height: 1.5;
  background-color: #fff;
  border-bottom: 1px solid #dcdfe6;

  &:empty {
    display: none;
  }

  .lv__header-title {
    display: inline-block;
    padding: 0 1em 0 0;
    margin: 0 1em 0 0;
    font-size: 16px;
    border-right: 1px solid #dcdfe6;
  }
  .lv__header-breadcrumb {
    display: inline-block;
    vertical-align: middle;
  }
}
</style>
