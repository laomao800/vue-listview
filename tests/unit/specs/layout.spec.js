import { mount } from '@vue/test-utils'
import Listview from '@/index.js'
import ListviewHeader from '@/components/listview-header'

describe('layout', () => {
  describe('listview header', () => {
    it('empty', () => {
      const wrapper = mount(ListviewHeader)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('title', () => {
      const wrapper = mount(ListviewHeader, {
        propsData: {
          title: 'headerTitle'
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('nav', () => {
      const wrapper = mount(ListviewHeader, {
        propsData: {
          nav: ['home', { text: 'list', to: '/list' }]
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('title & nav', () => {
      const wrapper = mount(ListviewHeader, {
        propsData: {
          title: 'headerTitle',
          nav: [{ text: 'home', to: '/home' }, { text: 'list', to: '/list' }]
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('height', () => {
    it('specify height', () => {
      const wrapper = mount(Listview, {
        propsData: {
          autoload: false,
          usePage: false,
          height: 500
        }
      })
      expect(wrapper.element.style.height).toBe('500px')
    })
  })

  describe('contentMessage', () => {
    it('string', () => {
      const wrapper = mount(Listview, {
        propsData: {
          contentMessage: 'message text'
        }
      })
      expect(wrapper.vm.internalContentMessage).toEqual({
        type: null,
        icon: null,
        message: 'message text'
      })
    })

    it('object', () => {
      const wrapper = mount(Listview, {
        propsData: {
          contentMessage: {
            type: 'info',
            message: 'message text'
          }
        }
      })
      expect(wrapper.vm.internalContentMessage).toEqual({
        type: 'info',
        icon: 'el-icon-info',
        message: 'message text'
      })
    })
  })

  it('pager off', () => {
    const wrapper = mount(Listview, { propsData: { usePage: false } })
    expect(wrapper.vm.$el.querySelector('.listview__page')).toBe(null)
  })

  describe('row class name', () => {
    it('row string class', async () => {
      const wrapper = mount(Listview, {
        propsData: {
          tableProps: {
            rowClassName: 'row-view-class'
          }
        }
      })
      expect(
        wrapper.vm.$refs.contentTable.rowClassName({
          row: {},
          rowIndex: 1
        })
      ).toContain('row-view-class')
    })

    it('row func class', () => {
      const wrapper = mount(Listview, {
        propsData: {
          tableProps: {
            rowClassName: ({ rowIndex }) => 'row-' + rowIndex
          }
        }
      })
      expect(
        wrapper.vm.$refs.contentTable.rowClassName({
          row: {},
          rowIndex: 1
        })
      ).toContain('row-1')
    })
  })
})
