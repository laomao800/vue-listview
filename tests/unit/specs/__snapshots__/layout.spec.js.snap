// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout contentMinHeight 1`] = `
<div class="listview" style="height: 200px; min-height: inherit;">
  <listviewheader-stub title="" nav=""></listviewheader-stub>
  <div class="listview__main">
    <filterbar-stub filterbuttons="" filterfields="" filtermodel="[object Object]" filterbarfold="true" showfiltersearch="true" showfilterreset="true"><template></template> <template></template></filterbar-stub>
    <div>
      <div class="listview__content" style="height: 500px;">
        <eltable-stub data="" size="small" fit="true" stripe="true" border="true" showheader="true" rowclassname="function () { [native code] }" selectonindeterminate="true" style="width: 100%;" height="500">
          <!---->
          <eltablecolumn-stub type="selection" width="50" resizable="true" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        </eltable-stub>
      </div>
      <div class="listview__page">
        <elpagination-stub pagesize="20" total="0" pagercount="7" currentpage="1" layout="total, sizes, prev, pager, next, jumper" pagesizes="20,50,100" background="true"></elpagination-stub>
      </div>
    </div>
  </div>
</div>
`;

exports[`layout listview header empty 1`] = `undefined`;

exports[`layout listview header nav 1`] = `
<div class="listview__header">
  <!---->
  <elbreadcrumb-stub separator="/" separatorclass="" class="listview__breadcrumb">
    <elbreadcrumbitem-stub>
      home
    </elbreadcrumbitem-stub>
    <elbreadcrumbitem-stub to="/list">
      list
    </elbreadcrumbitem-stub>
  </elbreadcrumb-stub>
</div>
`;

exports[`layout listview header title & nav 1`] = `
<div class="listview__header">
  <h1 class="listview__title">
    headerTitle
  </h1>
  <elbreadcrumb-stub separator="/" separatorclass="" class="listview__breadcrumb">
    <elbreadcrumbitem-stub to="/home">
      home
    </elbreadcrumbitem-stub>
    <elbreadcrumbitem-stub to="/list">
      list
    </elbreadcrumbitem-stub>
  </elbreadcrumb-stub>
</div>
`;

exports[`layout listview header title 1`] = `
<div class="listview__header">
  <h1 class="listview__title">
    headerTitle
  </h1>
  <elbreadcrumb-stub separator="/" separatorclass="" class="listview__breadcrumb"></elbreadcrumb-stub>
</div>
`;

exports[`layout pager 1`] = `
<div class="listview">
  <listviewheader-stub title="" nav=""></listviewheader-stub>
  <div class="listview__main">
    <filterbar-stub filterbuttons="" filterfields="" filtermodel="[object Object]" filterbarfold="true" showfiltersearch="true" showfilterreset="true"><template></template> <template></template></filterbar-stub>
    <div>
      <div class="listview__content">
        <eltable-stub data="" size="small" fit="true" stripe="true" border="true" showheader="true" rowclassname="function () { [native code] }" selectonindeterminate="true" style="width: 100%;">
          <!---->
          <eltablecolumn-stub type="selection" width="50" resizable="true" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        </eltable-stub>
      </div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`layout specify height 1`] = `
<div class="listview">
  <listviewheader-stub title="" nav=""></listviewheader-stub>
  <div class="listview__main">
    <filterbar-stub filterbuttons="" filterfields="" filtermodel="[object Object]" filterbarfold="true" showfiltersearch="true" showfilterreset="true"><template></template> <template></template></filterbar-stub>
    <div>
      <div class="listview__content">
        <eltable-stub data="" size="small" fit="true" stripe="true" border="true" showheader="true" rowclassname="function () { [native code] }" selectonindeterminate="true" style="width: 100%;">
          <!---->
          <eltablecolumn-stub type="selection" width="50" resizable="true" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        </eltable-stub>
      </div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`layout specify height 2`] = `
<div class="listview" style="height: 500px; min-height: inherit;">
  <listviewheader-stub title="" nav=""></listviewheader-stub>
  <div class="listview__main">
    <filterbar-stub filterbuttons="" filterfields="" filtermodel="[object Object]" filterbarfold="true" showfiltersearch="true" showfilterreset="true"><template></template> <template></template></filterbar-stub>
    <div>
      <div class="listview__content">
        <eltable-stub data="" size="small" fit="true" stripe="true" border="true" showheader="true" rowclassname="function () { [native code] }" selectonindeterminate="true" style="width: 100%;">
          <!---->
          <eltablecolumn-stub type="selection" width="50" resizable="true" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        </eltable-stub>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
